app/controllers/admin/contacts_controller.rb

```diff
@@ -0,0 +1,27 @@
+class Admin::ContactsController < AuthenticatedAdminsController
+  def edit
+    @contact = Contact.find(params[:id])
+  end
+
+  def update
+    @form = ::ContactForm.new(Contact.find(params[:id]))
+
+    if @form.update(update_params)
+      success_message(title: t('.success'))
+      token = @form.model.authorization_request.jwt_api_entreprise
```

Comment:
Une authorization request peut ne pas avoir de jeton.

Pour éviter ce souci, je namespacerai ce controller dans la resource `tokens` =>

```ruby
resources :tokens do
  resources :contacts, only: %i[index edit update]
end
```

La route générée aura forcément un `token_id`, et on autorise donc l'update que si il y a un jeton.

Et en vrai, c'est un peu brainfuck d'architecturer tout ça en fonction des jetons, on devrait revoir l'UX pour architecturer ça autour des demandes d'autorisations (à voir avec le refonte UX ..?)

Reply:
Alors atm, cette page est accessible uniquement depuis un jeton que ce soit pour les admins ou les users donc le soucis est limité depuis l'interface. 

Je suis favorable à attendre la refonte (UX et modèle oui) histoire de pas ajouter au brainfuck même temporairement WDYT ?

Reply:
> Alors atm, cette page est accessible uniquement depuis un jeton

Raison de plus de faire ce que je propose (ça retire toute ambiguïté en lisant le code). Le micmac authorization request/jetons peut clairement attendre la refonte par contre.

Reply:
Je vois ce que tu veux dire et est-ce que tu préférerais pas un scope sur Contacts pour justement limiter le périmètre ?

Ce qui me dérange avec l'ID dans l'URL c'est que c'est pas très REST parce qu'inutile en terme d'accès à la ressource. Ca sert à rien de nester `/a/:x/b/:y`, si t'as déjà :y l'ID de "b" t'as pas besoin de l'ID de "a" pour la retrouver. 

Je préfère un scope et je trouve ça plus clair.

Reply:
Ça me va. 

J'ai juste tendance à suivre la philo rails (et là pour le coup c'est un cas typique de nested resources : https://guides.rubyonrails.org/routing.html#nested-resources )

L'avantage c'est que c'est crystal clear que c'est scopé par la resource mère, et on se retrouve du coup à vérifier, par design, que la resource mère existe (et du coup on évite les potentiels opérations/problèmes/vérifications de la resource mère dans le code du controller).

Reply:
Bah justement en suivant la philo Rails c'est exactement le [shallow nesting](https://guides.rubyonrails.org/routing.html#shallow-nesting) que je propose 😄 Y a un bon article plus haut dans la doc Rails sur les limites du nesting. 

> L'avantage c'est que c'est crystal clear que c'est scopé par la resource mère, et on se retrouve du coup à vérifier, par design, que la resource mère existe (et du coup on évite les potentiels opérations/problèmes/vérifications de la resource mère dans le code du controller). 

Ca je comprends pas ? Si je fais un call à `/tokens/foo/contacts/bar` le controller il va tout laisser passer dans les params et dans tous les cas il faut check en db que les ressources existent, mères, filles, les deux. Et c'est au choix. Je trouve ça beaucoup moins crystal clear qu'un scope `.only_with_token` sur `Contact` par exemple.

Reply:
La ligne sur laquelle on commente là peut fail, vu qu'un contact peut ne pas avoir d'authorization request (et donc de jeton). 

La justification "on ne fait ça que depuis une page de jeton" n'est pas traduite dans ce controller, ce qui est un problème de design pour moi. 

> Si je fais un call à /tokens/foo/contacts/bar le controller il va tout laisser passer dans les params et dans tous les cas il faut check en db que les ressources existent, mères, filles, les deux.

Ici tu fais ça:

```ruby
jeton = JwtAPIEntreprise.find(params[:token_id])
contact = jeton.contacts.find(params[:id])
```

Si le jeton n'existe pas ça lève une 404 automatiquement avant l'update.

Reply:
Si on n'avait pas la redirection sur le jeton clairement le shallow nesting est adapté. Mon point ici c'est l'existence de ce jeton.

Reply:
> La justification "on ne fait ça que depuis une page de jeton" n'est pas traduite dans ce controller, ce qui est un problème de design pour moi.

Pour moi la responsabilité du routing c'est l'identification de la ressource. Qu'ici il faille que la dite ressource soit rattachée à des jetons ou autre c'est pas dans la route qu'il faut gérer ça. En tout cas c'est pas ici que je regarde. On a 100% un problème de design mais il vient du fait que le modèle ait changé : on avait des modèles qui étaient directement liés et maintenant il se retrouve avec un modèle entre eux. Pour le moment on bricole l'articulation entre la donnée parce qu'on a pas du tout repensé l'affichage de la donnée qui lui n'a pas bougé. 

> Ici tu fais ça:
> 
> jeton = JwtAPIEntreprise.find(params[:token_id])
> contact = jeton.contacts.find(params[:id])
> Si le jeton n'existe pas ça lève une 404 automatiquement avant l'update.

Oui je suis d'accord que l'erreur peut être levée, bon en pratique tu suis un lien depuis un jeton, c'est ce que je disais c'est limité comme erreur. Mais je suis d'accord c'est moche je vais refacto. Mais tu disais plus haut qu'avec ta solution tu évitais ce code dans le contrôleur et je vois pas comment.

Reply:
> Si on n'avait pas la redirection sur le jeton clairement le shallow nesting est adapté. Mon point ici c'est l'existence de ce jeton. 

Tout ce que dit le shallow nesting c'est qu'en terme d'identification de ressource c'est inutile de nester deux ressources avec ID. Les collections ça fait sens, pour le listing par exemple ce qu'on fait. C'est exactement notre situation ici.

Reply:
Nester la route offre une solution hard (si t'as pas de jeton tu peux juste pas accéder à l'action) + traduit directement qu'on ne peut pas accéder à la ressource en lisant la route. 

Dans les 2 cas il faut, à mon sens, récupérer la resource contact à travers le jeton si on veut traduire ce comportement de "n'accéder à la resource qu'à travers le jeton".

> Oui je suis d'accord que l'erreur peut être levée, bon en pratique tu suis un lien depuis un jeton, c'est ce que je disais c'est limité comme erreur.

Le problème de ce raisonnement c'est que tu limites une action de controller qu'avec la vue, hors la vue ne devrait pas "limiter" les actions du controller (vu qu'on peut forger des requêtes).

In fine je proposais la nested route car je trouvais ça plus logique de manager l'intégralité des contacts à travers les jetons (et par la suite des authorizations requests), mais as u ouiche du moment que la resource est retrouvé dans le scope du modèle mère.

Reply:
> Nester la route offre une solution hard (si t'as pas de jeton tu peux juste pas accéder à l'action) 

Si j'ai pas de jeton, je peux quand même accéder à l'action en tapant sur `/tokens/pasdeid/contacts/<id>`, et dans l'action il y a vérification que les records existent... Anyway je pense pas que ce soit très important. J'ai proposé un scope comme discuté.